py4cast/lightning.py

```diff
@@ -215,6 +225,26 @@ def __init__(self, hparams: ArLightningHyperParam, *args, **kwargs):
         self.psd_plot_metric = MetricPSDK(save_path, pred_step=max_pred_step)
         self.acc_metric = MetricACC(self.hparams["hparams"].dataset_info)
 
+    @property
+    def dtype(self):
+        """
+        Return the appropriate torch dtype for the desired precision in hparams.
+        """
+        return self.str_to_dtype[self.hparams["hparams"].precision]
```

Comment:
I thought about it but it would require that a trainer instance is always attached to the model even for inference and thus break the (official, per doc) first and simple way to do inference with an "unattached" lightning instance, see here: https://lightning.ai/docs/pytorch/stable/deploy/production_basic.html

I tried your suggestion and it yields an Exception when we infer using the ligtning module:
```python
    dl_settings = TorchDataloaderSettings(batch_size=1)
    infer_loader = infer_ds.torch_dataloader(dl_settings)
    preds = lightning_module(next(iter(infer_loader)))
 ```
 
 ```python
   File "/home/guibertf/py4cast/bin/inference.py", line 56, in <module>
    preds = lightning_module(next(iter(infer_loader)))
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
  File "/opt/conda/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
  File "/home/guibertf/py4cast/py4cast/lightning.py", line 540, in forward
    return self.common_step(x, inference=True)[0]
  File "/home/guibertf/py4cast/py4cast/lightning.py", line 354, in common_step
    with torch.cuda.amp.autocast(dtype=self.dtype):
  File "/home/guibertf/py4cast/py4cast/lightning.py", line 234, in dtype
    return self.str_to_dtype[self.trainer.precision]
  File "/opt/conda/lib/python3.10/site-packages/pytorch_lightning/core/module.py", line 218, in trainer
    raise RuntimeError(f"{self.__class__.__qualname__} is not attached to a `Trainer`.")
RuntimeError: AutoRegressiveLightning is not attached to a `Trainer`.
```

Also, as it is the precision is a single cli argument and value cascaded to both the Trainer and the lightning module so ihmo there is no risk of mismatch. Also, I don"t want tostore and force the training precision at inference time, since it is not uncommon to train in bf16 and infer in float32.